clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+import org.apache.kafka.clients.consumer.internals.events.BackgroundEvent;
+import org.apache.kafka.clients.consumer.internals.events.RebalanceCallbackEvent;
+import org.apache.kafka.clients.consumer.internals.events.RevokePartitionsEvent;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatResponseData.Assignment;
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatResponseData.TopicPartitions;
+import org.apache.kafka.clients.consumer.internals.Utils.TopicPartitionComparator;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Timer;
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.BlockingQueue;
+import java.util.stream.Collectors;
+
+/**
+ * {@code MemberAssignmentReconciler} works with {@link MembershipManager} to a) determine, and b) modify the
+ * current set of assigned {@link TopicPartition partitions} via {@link SubscriptionState}. Reconciliation is a
+ * two-part process, the first being a revocation of partitions, followed by assignment of partitions. Each of the two
+ * steps may result in one of the following:
+ *
+ * <ul>
+ *     <li>
+ *         {@link ReconciliationResult#NO_CHANGE}: no changes were made to the set of partitions.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#IN_PROGRESS}: changes to the assignment have started. In practice this means
+ *         that the appropriate {@link ConsumerRebalanceListener} callback method is being invoked.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#COMPLETED}: the {@link ConsumerRebalanceListener} callback method was made and
+ *         the changes were applied locally.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#EXPIRED}: something happened to cause the operation to modify the assigned set
+ *         of partitions. This could be caused by a {@link ConsumerRebalanceListener} callback method that takes too
+ *         long to execute, interruption with the consumer group coordinator, or other reasons.
+ *     </li>
+ * </ul>
+ *
+ * The comparison against the {@link SubscriptionState#assignedPartitions() current set of assigned partitions} and
+ * the {@link Assignment#assignedTopicPartitions() target set of assigned partitions} is performed by essentially
+ * <em>flattening</em> the respective entries into two sets of {@link org.apache.kafka.common.TopicPartition partitons}
+ * which are then compared using basic {@link Set} comparisons.
+ */
+public class MemberAssignmentReconciler {
```

Comment:
One missing piece would be the heartbeat - the heartbeat needs to be sent out upon revoking partitions. I think we can leave a note here or implement this in this PR.

Reply:
The `revoke()` and `assign()` methods return a `ReconciliationResult` to let the caller know the operation completed. In that case, the caller would work with the `HeartbeatRequestManager` so that it includes the updated partition set in its next request.